pkg/controller/groups/deploytokens/controller.go

```diff
@@ -0,0 +1,199 @@
+func SetupDeployToken(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.DeployTokenKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.DeployToken{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.DeployTokenGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), newGitlabClientFn: groups.NewDeployTokenClient}),
+			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube              client.Client
+	newGitlabClientFn func(cfg clients.Config) groups.DeployTokenClient
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.DeployToken)
+	if !ok {
+		return nil, errors.New(errNotDeployToken)
+	}
+	cfg, err := clients.GetConfig(ctx, c.kube, cr)
+	if err != nil {
+		return nil, err
+	}
+	return &external{kube: c.kube, client: c.newGitlabClientFn(*cfg)}, nil
+}
+
+type external struct {
+	kube   client.Client
+	client groups.DeployTokenClient
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.DeployToken)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotDeployToken)
+	}
+
+	externalName := meta.GetExternalName(cr)
+	if externalName == "" {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	groupDeployTokenID, err := strconv.Atoi(externalName)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.New(errNotDeployToken)
+	}
+
+	listGroupDeployTokensOptions := gitlab.ListGroupDeployTokensOptions{}
+	deploytokenArr, _, err := e.client.ListGroupDeployTokens(*cr.Spec.ForProvider.GroupID, &listGroupDeployTokensOptions)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(groups.IsErrorGroupDeployTokenNotFound, err), errGetFailed)
+	}
+
+	dt := findDeployToken(groupDeployTokenID, deploytokenArr)
+
+	if dt == nil {
+		return managed.ExternalObservation{}, nil
+	}
+
+	current := cr.Spec.ForProvider.DeepCopy()
+	lateInitializeGroupDeployToken(&cr.Spec.ForProvider, dt)
+
+	cr.Status.AtProvider = v1alpha1.DeployTokenObservation{}
+	cr.Status.SetConditions(xpv1.Available())
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        true,
+		ResourceLateInitialized: !cmp.Equal(current, &cr.Spec.ForProvider),
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.DeployToken)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotDeployToken)
+	}
+
+	dt, _, err := e.client.CreateGroupDeployToken(
+		*cr.Spec.ForProvider.GroupID,
```

Comment:
I get a nil pointer exception if I leave the group field empty (which is a optional field)

Reply:
Implemented in #53